server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledIterator.java

```diff
@@ -49,84 +51,49 @@ public static <T> void run(Iterator<T> iterator, BiConsumer<Releasable, T> itemC
         if (maxConcurrency <= 0) {
             throw new IllegalArgumentException("maxConcurrency must be positive");
         }
-        this.permits = new Semaphore(maxConcurrency);
+        this.permits = new AtomicInteger(maxConcurrency);
         this.refs = AbstractRefCounted.of(onCompletion);
     }
 
     private void run() {
-        while (permits.tryAcquire()) {
+        do {
             final T item;
-            synchronized (iterator) {
-                if (iterator.hasNext()) {
-                    item = iterator.next();
-                } else {
-                    permits.release();
-                    return;
-                }
+            if (iterator.hasNext()) {
+                item = iterator.next();
+            } else {
+                return;
             }
-            try (var itemRefs = new ItemRefCounted()) {
-                itemRefs.mustIncRef();
-                itemConsumer.accept(Releasables.releaseOnce(itemRefs::decRef), item);
+            try {
+                refs.mustIncRef();
+                itemConsumer.accept(Releasables.releaseOnce(itemReleasable), item);
             } catch (Exception e) {
                 logger.error(Strings.format("exception when processing [%s] with [%s]", item, itemConsumer), e);
                 assert false : e;
             }
-        }
+        } while (permits.decrementAndGet() > 0);
     }
 
     @Override
     public void close() {
         refs.decRef();
     }
 
-    // A RefCounted for a single item, including protection against calling back into run() if it's created and closed within a single
-    // invocation of run().
-    private class ItemRefCounted extends AbstractRefCounted implements Releasable {
-        private boolean isRecursive = true;
-
-        ItemRefCounted() {
-            refs.mustIncRef();
-        }
-
-        @Override
-        protected void closeInternal() {
-            permits.release();
-            try {
-                // Someone must now pick up the next item. Here we might be called from the run() invocation which started processing the
-                // just-completed item (via close() -> decRef()) if that item's processing didn't fork or all its forked tasks finished
-                // first. If so, there's no need to call run() here, we can just return and the next iteration of the run() loop will
-                // continue the processing; moreover calling run() in this situation could lead to a stack overflow. However if we're not
-                // within that run() invocation then ...
-                if (isRecursive() == false) {
-                    // ... we're not within any other run() invocation either, so it's safe (and necessary) to call run() here.
-                    run();
-                }
-            } finally {
-                refs.decRef();
+    private void onItemRelease() {
+        try {
+            if (permits.getAndIncrement() == 0) {
```

Comment:
This neatly removes the need for special recursion-detection: `permits` remains strictly positive until the very last line of `run()`, so the only way we can take this `if()` branch is if `run()` has finished its work.